### ISSUE TITLE
Tensor descriptor suppport investigation

### ISSUE THREAD

**Issue by mitekoth** [facebookincubator/triton-shared#7]
I was investigating the lowering support when using tensor descriptors instead of block pointers for the triton kernels as block pointers are now being deprecated and here are some of my observations:
- We can replace block pointers with tensor descriptors with minimal changes by adding [RewriteTensorDescriptorToPointer](https://github.com/facebookincubator/triton-shared/blob/942486775f678d2961efe0fcdd3f37a892d4e2cc/backend/compiler.py#L256) which is a triton pass. This ensures functionality.
- But the IR generated is different from the block pointer approach as the [rewrite tensor descriptor pass](https://github.com/sjw36/triton/blob/main/lib/Dialect/Triton/Transforms/RewriteTensorDescriptorToPointer.cpp) is creating masked loads all the time.

So we can adopt two different approaches:
- Add a canonicalizer pass to identify bounded loads and remove the masks.
- Add explicit lowering to tt.make_tensor_descriptor ops similar to triton::MakeTensorPtrOp and tt.AdvanceOp

Attaching two equivalent IRs where the only difference in input triton kernel was using block pointers (or) tensor descriptors.
1. Generated from triton kernel with block pointers - 
[tensor_desc_tritonshared_lowering.log](https://github.com/user-attachments/files/26030878/tensor_desc_tritonshared_lowering.log)

2. Generated from triton kernel with tensor descriptors - 
[block_ptr_tritonshared_lowering.log](https://github.com/user-attachments/files/26030884/block_ptr_tritonshared_lowering.log)

**Comment by enjustli** [comment]
I think we should directly handle `tt.make_tensor_descriptor`, just like the downstream passes for `NV`/`AMD` do. 
`RewriteTensorDescriptorToPointer` is intended only for hardware that does not support `tt.make_tensor_descriptor`. 
Based on my past experience, using the `canonicalizer` for boundary analysis tends to introduce many additional constraints; it’s better to handle `tt.make_tensor_descriptor` directly.

**Comment by nhat-nguyen** [comment]
@mitekoth Thanks for sharing the logs. Can you share what your block pointer in triton (not the IR) looks like? Are you using `padding_option` for your block pointers?

Most likely it's because the block pointer version in your kernel doesn't have `padding_option` which means there's no mask.

In tensor descriptor, the padding seems to be a requirement. It defaults to 0 if you don't provide anything when calling `make_tensor_descriptor`:
- https://sourcegraph.com/-/editor?remote_url=https%3A%2F%2Fgithub.com%2Ftriton-lang%2Ftriton.git&branch=main&file=include%2Ftriton%2FDialect%2FTriton%2FIR%2FTritonAttrDefs.td&editor=VSCode&version=2.4.0&start_row=41&start_col=4&end_row=41&end_col=24
- https://sourcegraph.com/-/editor?remote_url=https%3A%2F%2Fgithub.com%2Ftriton-lang%2Ftriton.git&branch=main&file=include%2Ftriton%2FDialect%2FTriton%2FIR%2FTritonOps.td&editor=VSCode&version=2.4.0&start_row=989&start_col=0&end_row=990&end_col=0

If we intentionally pass `padding_option=None`, the triton frontend will crash.

Tested with the following kernel:

```python
import torch
import triton
import triton.language as tl


@triton.jit
def copy_with_tensor_desc(
    in_ptr,
    out_ptr,
    M,
    N,
    stride_m,
    stride_n,
    BLOCK_M: tl.constexpr,
    BLOCK_N: tl.constexpr,
):
    """
    Simple Triton kernel demonstrating tensor descriptor load and store
    using make_tensor_descriptor (device-side API).
    """
    # Get program ID for this block
    pid_m = tl.program_id(0)
    pid_n = tl.program_id(1)
    
    # Calculate offsets for this block
    offset_m = pid_m * BLOCK_M
    offset_n = pid_n * BLOCK_N
    
    # Create tensor descriptors using make_tensor_descriptor (device-side)
    in_desc = tl.make_tensor_descriptor(
        in_ptr,
        shape=[M, N],
        strides=[stride_m, stride_n],
        block_shape=[BLOCK_M, BLOCK_N],
        padding_option=None, # None will crash, default to 0 if not provided
    )
    
    out_desc = tl.make_tensor_descriptor(
        out_ptr,
        shape=[M, N],
        strides=[stride_m, stride_n],
        block_shape=[BLOCK_M, BLOCK_N],
    )
    
    # Load a tile from input using descriptor_load
    tile = tl.load_tensor_descriptor(in_desc, [offset_m, offset_n])
    
    # Store the tile to output using descriptor_store
    tl.store_tensor_descriptor(out_desc, [offset_m, offset_n], tile)


# Test the kernel
M, N = 256, 256
BLOCK_M, BLOCK_N = 64, 64

input_tensor = torch.randn(M, N, device="cuda", dtype=torch.float32)
output_tensor = torch.empty_like(input_tensor)

grid = (triton.cdiv(M, BLOCK_M), triton.cdiv(N, BLOCK_N))

copy_with_tensor_desc[grid](
    input_tensor,
    output_tensor,
    M, N,
    input_tensor.stride(0),
    input_tensor.stride(1),
    BLOCK_M=BLOCK_M,
    BLOCK_N=BLOCK_N,
)

# Verify the copy is correct
if torch.allclose(input_tensor, output_tensor):
    print("✓ Tensor descriptor copy successful!")
    print(f"  Input shape: {input_tensor.shape}")
    print(f"  Output shape: {output_tensor.shape}")
    print(f"  Max difference: {(input_tensor - output_tensor).abs().max().item()}")
else:
    print("✗ Tensor descriptor copy failed!")
    print(f"  Max difference: {(input_tensor - output_tensor).abs().max().item()}")
```

**Comment by mitekoth** [comment]
@nhat-nguyen Yes that's correct. The triton kernel that I was experimenting with doesn't have `padding_option` set. If that's set, the codegen may be similar to the `tensor_descriptor` default scenario now.

In any case, to effectively replace `block_pointers` in kernels which don't require padding, we should add `tensor_descriptor` lowering support in triton_shared itself instead of relying on `RewriteTensorDescriptorToPointer` as @enjustli has also mentioned.

If that sounds good to you, I can get started on an initial implementation for the lowering.

**Comment by nhat-nguyen** [comment]
> [@nhat-nguyen](https://github.com/nhat-nguyen) Yes that's correct. The triton kernel that I was experimenting with doesn't have `padding_option` set. If that's set, the codegen may be similar to the `tensor_descriptor` default scenario now.
> 
> In any case, to effectively replace `block_pointers` in kernels which don't require padding, we should add `tensor_descriptor` lowering support in triton_shared itself instead of relying on `RewriteTensorDescriptorToPointer` as [@enjustli](https://github.com/enjustli) has also mentioned.
> 
> If that sounds good to you, I can get started on an initial implementation for the lowering.

Yeah I agree this change in masking behavior is noteworthy. We're doing some internal investigations too. Hopefully we will be able to spin up a conversation soon to understand if upstream triton should support tensor descriptor with no masks.

**Comment by enjustli** [comment]
i submitted a PR: https://github.com/facebookincubator/triton-shared/pull/25
